### PR TITLE
Implement Smart Pointer Creation With Default Initialization

### LINKED ISSUE #48: P1020R1 Smart Pointer Creation With Default Initialization
[P1020R1](https://wg21.link/P1020R1) Smart Pointer Creation With Default Initialization

[P1973R1](https://wg21.link/P1973R1) Renaming `default_init` To `for_overwrite`

Final feature-test macro in WG21-N4861:

`#define __cpp_lib_smart_ptr_for_overwrite 202002L`

### PR DESCRIPTION
**Pull request by AdamBucior**
Resolves #48. Based on #778.

### CODE REVIEW

File: stl/inc/memory [hunk 1]
```diff
@@ -2878,6 +2952,36 @@ _NODISCARD enable_if_t<is_bounded_array_v<_Ty>, shared_ptr<_Ty>> make_shared(con
     _Ret._Set_ptr_rep_and_enable_shared(_Rx->_Storage._Value, _Rx);
     return _Ret;
 }
+
+// FUNCTION TEMPLATE make_shared_for_overwrite
+template <class _Ty>
+_NODISCARD enable_if_t<!is_unbounded_array_v<_Ty>, shared_ptr<_Ty>> make_shared_for_overwrite() {
+    shared_ptr<_Ty> _Ret;
+    if constexpr (!is_array_v<_Ty>) {
```
Comment (miscco) [inline stl/inc/memory:2960]: I would prefere if you swap the blocks. It is always easier to follow `meow` rather than `!meow`

File: stl/inc/memory [hunk 2]
```diff
@@ -2990,6 +3094,54 @@ _NODISCARD enable_if_t<is_bounded_array_v<_Ty>, shared_ptr<_Ty>> allocate_shared
     _Ret._Set_ptr_rep_and_enable_shared(_Ptr, _Unfancy(_Constructor._Release()));
     return _Ret;
 }
+
+// FUNCTION TEMPLATE allocate_shared_for_overwrite
+template <class _Ty, class _Alloc>
+_NODISCARD enable_if_t<!is_unbounded_array_v<_Ty>, shared_ptr<_Ty>> allocate_shared_for_overwrite(const _Alloc& _Al) {
+    shared_ptr<_Ty> _Ret;
+    if constexpr (!is_array_v<_Ty>) {
```
Comment (miscco) [inline stl/inc/memory:3102]: Same here and throughout

File: stl/inc/memory (hunk 1 above)
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory (hunk 2 above)
Comment (AdamBucior) [reply stl/inc/memory]: Done

**Comment by StephanTLavavej** [comment]
I pushed a commit (deleting an extra newline) with the `--author=` option, so when we merge this PR, it will be displayed as co-authored by @Weheineman. Thanks for picking up this work!

**Comment by StephanTLavavej** [comment]
Status update: Planning to review this on Thursday, Oct 8. Apologies for the delay!

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Thanks! This looks good, my main concern is with a potential corner case in overload resolution that should be avoidable with different metaprogramming. Please let me know if you'd like me to explain further, or I can revise the PR if you want - your preference.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+template <typename T>
+struct unique_is_for_overwritable<T, void_t<decltype(make_unique_for_overwrite<T>())>> : true_type {};
+
+template <typename T>
+constexpr bool unique_is_for_overwritable_v = unique_is_for_overwritable<T>::value;
+
+struct DefaultInitializableInt {
+    int value;
+    DefaultInitializableInt() : value(106) {}
```
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:75]: Could you turn the magic number into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        assert(descendingAddressBuffer[i - 1] > descendingAddressBuffer[i]);
+    }
+
+    descendingAddressBuffer.clear();
+}
+
+void assert_uninitialized(void* p, size_t size) {
+    unsigned char* chPtr = reinterpret_cast<unsigned char*>(p);
+    for (size_t offset = 0; offset < size; ++offset) {
+        assert(*(chPtr + offset) == 0xEE);
```
[the remark this answers is not in the document]
Comment (TylerMSFT) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:122]: Could you turn the magic number 0xEE into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    } else {
+        throw bad_alloc();
+    }
+}
+
+void* operator new(size_t size, const nothrow_t&) noexcept {
+    void* const result = malloc(size == 0 ? 1 : size);
+    ++allocationCount;
+    if (result) {
+        memset(result, 0xEE, size);
```
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:38]: Could you turn the magic number 0xEE into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    } else {
+        throw bad_alloc();
+    }
+}
+
+void* operator new(size_t size, align_val_t align, const nothrow_t&) noexcept {
+    void* const result = ::_aligned_malloc(size, static_cast<size_t>(align));
+    ++allocationCount;
+    if (result) {
+        memset(result, 0xEE, size);
```
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:58]: Could you turn the magic number 0xEE into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: stl/inc/memory [hunk 7]
```diff
@@ -2172,6 +2190,10 @@ public:
         _Construct_in_place(_Storage._Value, _STD forward<_Types>(_Args)...);
     }
 
+    explicit _Ref_count_obj2(_For_overwrite_tag) : _Ref_count_base() {
+        _Construct_in_place_for_overwrite(_Storage._Value);
+    }
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/memory:2195]: @cdacamar noted that we can deal with the variadic `_Types` without introducing a wrapper class, by using fold expressions. That is, `(is_same_v<_Types, _For_overwrite_tag> && ...)` (I believe that's the syntax) combined with `sizeof...(_Types) == 1` should be sufficient.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp (hunk 6 above)
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp (hunk 5 above)
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp (hunk 3 above)
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: stl/inc/memory (hunk 7 above)
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory [hunk 8]
```diff
@@ -2387,6 +2433,10 @@ public:
         _Uninitialized_fill_multidimensional_n(_Get_ptr(), _Count, _Val);
     }
 
+    explicit _Ref_count_unbounded_array(const size_t _Count, _For_overwrite_tag) : _Ref_count_base() {
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory [hunk 9]
```diff
@@ -2878,6 +2952,35 @@ _NODISCARD enable_if_t<is_bounded_array_v<_Ty>, shared_ptr<_Ty>> make_shared(con
...
+    } else {
+        // make a shared_ptr to non-array object
+        const auto _Rx = new _Ref_count_obj2<_Ty>(_For_overwrite_tag{});
+        _Ret._Set_ptr_rep_and_enable_shared(_STD addressof(_Rx->_Storage._Value), _Rx);
+    }
+    return _Ret;
+}
+
+template <class _Ty>
+_NODISCARD enable_if_t<is_unbounded_array_v<_Ty>, shared_ptr<_Ty>> make_shared_for_overwrite(size_t _Count) {
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory [hunk 10]
```diff
@@ -2878,6 +2952,35 @@ _NODISCARD enable_if_t<is_bounded_array_v<_Ty>, shared_ptr<_Ty>> make_shared(con
...
+    // make a shared_ptr to an unbounded array
+    using _Refc    = _Ref_count_unbounded_array<_Ty>;
+    const auto _Rx = _Allocate_flexible_array<_Refc>(_Count);
+    _Global_delete_guard<_Refc> _Guard{_Rx};
+    ::new (static_cast<void*>(_Rx)) _Refc(_Count, _For_overwrite_tag{});
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+    return sp;
+}
+
+template <class T, class... Args>
+void test_make_init_destruct_order(Args&&... vals) {
+    try {
+        shared_ptr<T> sp = make_shared_for_overwrite<T>(forward<Args>(vals)...);
+        assert_shared_use_get(sp);
+    } catch (const runtime_error& exc) {
+        assert(exc.what() == "Can't create more ReportAddress objects."sv);
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:153]: Done

File: stl/inc/memory [hunk 12]
```diff
@@ -3450,6 +3601,23 @@ _NODISCARD unique_ptr<_Ty> make_unique(size_t _Size) { // make a unique_ptr
...
 
+#if _HAS_CXX20
+// FUNCTION TEMPLATE make_unique_for_overwrite
+template <class _Ty, enable_if_t<!is_array_v<_Ty>, int> = 0>
+_NODISCARD unique_ptr<_Ty> make_unique_for_overwrite() { // make a unique_ptr with default initialization
+    return unique_ptr<_Ty>(new _Ty);
+}
+
+template <class _Ty, enable_if_t<is_array_v<_Ty> && extent_v<_Ty> == 0, int> = 0>
+_NODISCARD unique_ptr<_Ty> make_unique_for_overwrite(size_t _Size) { // make a unique_ptr with default initialization
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+template <typename T>
+constexpr bool unique_is_for_overwritable_v = unique_is_for_overwritable<T>::value;
+
+struct DefaultInitializableInt {
+    int value;
+    DefaultInitializableInt() : value(106) {}
+};
+
+struct alignas(32) HighlyAligned {
+    uint64_t a;
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:85]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+template <class T>
+void assert_shared_use_get(const shared_ptr<T>& sp) {
+    assert(sp.use_count() == 1);
+    assert(sp.get() != nullptr);
+}
+
+template <class T, class... Args>
+shared_ptr<T> make_shared_for_overwrite_assert(Args&&... vals) {
+    size_t count     = allocationCount;
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+    shared_ptr<T> sp = make_shared_for_overwrite<T>(forward<Args>(vals)...);
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:141]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+    assert_shared_use_get(sp);
+    assert(count + 1 == allocationCount);
+    return sp;
+}
+
+template <class T, class... Args>
+void test_make_init_destruct_order(Args&&... vals) {
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+shared_ptr<T> allocate_shared_for_overwrite_assert(Args&&... vals) {
+    size_t aCount    = allocationCount;
+    shared_ptr<T> sp = allocate_shared_for_overwrite<T>(forward<Args>(vals)...);
+    assert_shared_use_get(sp);
+    assert(aCount + 1 == allocationCount);
+    return sp;
+}
+
+template <class T, class... Args>
+void test_allocate_init_destruct_order(Args&&... vals) {
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+    auto p0 = allocate_shared_for_overwrite_assert<int>(a0);
+    assert_uninitialized(addressof(*p0), sizeof(int));
+
+    allocator<DefaultInitializableInt> a1{};
+    auto p1 = allocate_shared_for_overwrite_assert<DefaultInitializableInt>(a1);
+    assert(p1->value == 106);
+
+    allocator<HighlyAligned> a2{};
+    auto p2 = allocate_shared_for_overwrite_assert<HighlyAligned>(a2);
+    assert_uninitialized(addressof(*p2), sizeof(HighlyAligned));
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: stl/inc/memory [hunk 19]
```diff
@@ -3450,6 +3601,23 @@ _NODISCARD unique_ptr<_Ty> make_unique(size_t _Size) { // make a unique_ptr
...
 void make_unique(_Types&&...) = delete;
 
+#if _HAS_CXX20
+// FUNCTION TEMPLATE make_unique_for_overwrite
+template <class _Ty, enable_if_t<!is_array_v<_Ty>, int> = 0>
+_NODISCARD unique_ptr<_Ty> make_unique_for_overwrite() { // make a unique_ptr with default initialization
+    return unique_ptr<_Ty>(new _Ty);
+}
+
+template <class _Ty, enable_if_t<is_array_v<_Ty> && extent_v<_Ty> == 0, int> = 0>
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory [hunk 20]
```diff
@@ -2990,6 +3093,54 @@ _NODISCARD enable_if_t<is_bounded_array_v<_Ty>, shared_ptr<_Ty>> allocate_shared
...
+_NODISCARD enable_if_t<!is_unbounded_array_v<_Ty>, shared_ptr<_Ty>> allocate_shared_for_overwrite(const _Alloc& _Al) {
+    shared_ptr<_Ty> _Ret;
+    if constexpr (is_array_v<_Ty>) {
+        // make a shared_ptr to a bounded array
+        using _Refc    = _Ref_count_bounded_array_alloc<remove_cv_t<_Ty>, _Alloc>;
+        using _Alblock = _Rebind_alloc_t<_Alloc, _Refc>;
+        _Alblock _Rebound(_Al);
+        _Alloc_construct_ptr _Constructor{_Rebound};
+        _Constructor._Allocate();
+        ::new (static_cast<void*>(_Unfancy(_Constructor._Ptr))) _Refc(_Al, _For_overwrite_tag{});
```
[the remark this answers is not in the document]
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+#include <stdexcept>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+using namespace std;
+
+constexpr int uninitializedValue = 0xEE;
+constexpr int defaultValue       = 106;
```
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:19]: The meaning of 'defaultValue` is very broad. Perhaps: initializedMemoryValue?

Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Renamed to `initializedValue`

**Comment by AdamBucior** [comment]
It's ready for another review.

**Comment by TylerMSFT** [comment]
@AdamBucior, thank you for addressing my feedback. I'm good with your changes. I defer to Stephan for the remaining review.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
This looks great - thanks for the thorough and highly detailed work. (I noticed and appreciated the improvements like using `remove_cvref_t` on perfectly forwarded `_Types` to handle const lvalues, and unifying additional repetitive code. Combining 3 overloads into 1, for both the `make` and `allocate` families, is awesome.)

I pushed a merge with `master` and a small set of superficial cleanups, after verifying that the test is still passing:

* When granting friendship to `_Make_shared_unbounded_array` and `_Allocate_shared_unbounded_array`, drop `const` on the value parameter `size_t _Count`. (We conventionally avoid top-level `const` on value parameters in function declarations, since it has no meaning to callers and is immediately ignored by the language. We are so picky about `const`! :joy_cat:)
* Your `_Make_shared_unbounded_array`/`_Allocate_shared_unbounded_array` refactoring has a bonus: we no longer need to grant friendship to several `make_shared` etc. overloads.
* When functions take `const _Arg&`, we don't need to apply `remove_cvref_t` to `_Arg` (unless we were worried about `_Arg` being passed as an explicit template argument, instead of being deduced, which is not a concern here).
* Remove a leftover comment after the `allocate_shared` refactoring.
* Use `is_bounded_array_v` in `make_unique_for_overwrite` for consistency.

I'll find a second maintainer to review and then we can merge this! :rocket:

File: stl/inc/memory [hunk 22]
```diff
@@ -2351,6 +2369,25 @@ void _Uninitialized_value_construct_multidimensional_n(_Ty* const _Out, const si
...
+            for (size_t& _Idx = _Guard._Index; _Idx < _Size; ++_Idx) {
+                _Uninitialized_default_construct_multidimensional_n(_Out[_Idx], extent_v<_Ty>);
+            }
+            _Guard._Target = nullptr;
+        } else {
+            _Uninitialized_rev_destroying_backout _Backout{_Out};
+            for (size_t _Idx = 0; _Idx < _Size; ++_Idx) {
+                _Backout._Emplace_back_for_overwrite();
+            }
+            _Backout._Release();
```
Comment (miscco) [inline stl/inc/memory:2386]: Should we simplify this to
```suggestion
            _Uninitialized_default_construct_n(_Out, _Size); 
```

File: stl/inc/xutility [hunk 23]
```diff
@@ -126,6 +126,12 @@ void _Construct_in_place(_Ty& _Obj, _Types&&... _Args) noexcept(is_nothrow_const
         _Ty(_STD forward<_Types>(_Args)...);
 }
 
+// FUNCTION TEMPLATE _Construct_in_place_for_overwrite
+template <class _Ty>
+void _Construct_in_place_for_overwrite(_Ty& _Obj) noexcept(is_nothrow_default_constructible_v<_Ty>) {
+    ::new (const_cast<void*>(static_cast<const volatile void*>(_STD addressof(_Obj)))) _Ty;
```
Comment (miscco) [inline stl/inc/xutility:132]: I think we should use `_Voidify_iter`
```suggestion
    ::new (_Voidify_iter(_STD addressof(_Obj))) _Ty;
```

File: stl/inc/xutility [hunk 24]
```diff
@@ -126,6 +126,12 @@ void _Construct_in_place(_Ty& _Obj, _Types&&... _Args) noexcept(is_nothrow_const
         _Ty(_STD forward<_Types>(_Args)...);
 }
 
+// FUNCTION TEMPLATE _Construct_in_place_for_overwrite
+template <class _Ty>
+void _Construct_in_place_for_overwrite(_Ty& _Obj) noexcept(is_nothrow_default_constructible_v<_Ty>) {
```
Comment (miscco) [inline stl/inc/xutility:131]: I really do not like the name, maybe `_Default_construct_in_place`?

If you do change it, would you mind using it for the `uninitialized_default_construct{_n}` family

File: stl/inc/memory [hunk 25]
```diff
@@ -2351,6 +2369,25 @@ void _Uninitialized_value_construct_multidimensional_n(_Ty* const _Out, const si
     }
 }
 
+template <class _Ty>
+void _Uninitialized_default_construct_multidimensional_n(_Ty* const _Out, const size_t _Size) {
+    if constexpr (!is_trivially_default_constructible_v<_Ty>) {
```
Comment (miscco) [inline stl/inc/memory:2374]: To be absolutely sure, even for the `meow_overwrite` we do not want to default initialize here?

Comment (AdamBucior) [reply stl/inc/memory:2374]: That's the whole point of `*_for_overwrite`

File: stl/inc/memory (hunk 22 above)
Comment (AdamBucior) [reply stl/inc/memory:2386]: There is no such function as `_Uninitialized_default_construct_n` and we can't use `uninitialized_default_construct_n` because it doesn't destroy elements in the reverse order.

File: stl/inc/xutility (hunk 23 above)
Comment (AdamBucior) [reply stl/inc/xutility]: Done

File: stl/inc/xutility (hunk 24 above)
Comment (AdamBucior) [reply stl/inc/xutility]: Done

File: stl/inc/memory (hunk 22 above)
Comment (miscco) [reply stl/inc/memory:2381]: Hm there is the ranges version (where I found a unneeded specialization that I forgot to remove so thanks ;)) 

Should we also destroy in backwards order there?

Comment (AdamBucior) [reply stl/inc/memory:2381]: `ranges::uninitialized_default_construct_n` needs to support forward iterators and destroying in reverse order with them would be hard.

File: stl/inc/memory (hunk 25 above)
Comment (CaseyCarter) [reply stl/inc/memory:2369]: For better or worse, there are many places in the STL where we assume that we can omit default-initialization of objects that are trivially default constructible. I'm not sure the resulting behavior is portably defined, but it has been working as intended on our implementation. We may need to make changes if and when compilers complain about this trick in `constexpr`, but that hasn't happened yet and in any case would need for more widespread changes than just this PR.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/yvals_core.h [hunk 26]
```diff
@@ -1208,6 +1210,7 @@
 #define __cpp_lib_remove_cvref                 201711L
 #define __cpp_lib_semaphore                    201907L
 #define __cpp_lib_shift                        201806L
+#define __cpp_lib_smart_ptr_for_overwrite      201811L
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/yvals_core.h]: Thanks for catching this!

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks - the latest changes look good. I was concerned about debug codegen impact from increased use of `_Voidify_iter` (as we would `addressof` twice in several cases), but it does centralize the obnoxious casts. So I believe a reasonable compromise is to add a simple throughput optimization, avoiding deref-and-`addressof` when the input is already a raw pointer, and I went ahead and pushed that. For precedent, we do the same thing in `_Unfancy` although without `if constexpr`:
https://github.com/microsoft/STL/blob/e7d5113376763268075d87dbad662a7cf4763a10/stl/inc/xstddef#L281-L290

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this feature, @AdamBucior and @Weheineman! This will help users achieve maximum performance with Standard smart pointers. :rocket:
